examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2218 @@
+        if (is_artificial)
+          return;
+
+        local_boundary_normal_map.clear();
+        cell_lumped_mass_matrix.reinit(dofs_per_cell, dofs_per_cell);
+        for (auto &matrix : cell_cij_matrix)
+          matrix.reinit(dofs_per_cell, dofs_per_cell);
+
+        const auto &fe_values = scratch.reinit(cell);
+
+        local_dof_indices.resize(dofs_per_cell);
+        cell->get_dof_indices(local_dof_indices);
+
+        std::transform(local_dof_indices.begin(),
+                       local_dof_indices.end(),
+                       local_dof_indices.begin(),
+                       [&](auto index) {
+                         return partitioner->global_to_local(index);
+                       });
+
+        /* We compute the local contributions for the  lumped mass
+         matrix entries m_i and and vectors c_ij */
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            const auto JxW = fe_values.JxW(q_point);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                const auto value_JxW = fe_values.shape_value(j, q_point) * JxW;
+                const auto grad_JxW  = fe_values.shape_grad(j, q_point) * JxW;
+
+                cell_lumped_mass_matrix(j, j) += value_JxW;
+
+                for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                  {
+                    const auto value = fe_values.shape_value(i, q_point);
+                    for (unsigned int d = 0; d < dim; ++d)
+                      cell_cij_matrix[d](i, j) += (value * grad_JxW)[d];
+
+                  } /* for i */
+              }     /* for j */
+          }         /* for q */
+
+        /* Now we have to compute the boundary normals. Note that the
+           following loop does not actually do much unless the faces of the
+           cell are actually faces on the boundary of the domain */
+        for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+          {
+            const auto face = cell->face(f);
+            const auto id   = face->boundary_id();
+
+            if (!face->at_boundary())
+              continue;
+
+            const auto &fe_face_values = scratch.reinit(cell, f);
+
+            const unsigned int n_face_q_points =
+              fe_face_values.get_quadrature().size();
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                if (!discretization->finite_element.has_support_on_face(j, f))
+                  continue;
+
+                /* Note that "normal" will only represent the contributions 
+                   from one of the faces in the support of the shape 
+                   function \phi_j. So we cannot normalize this local 
+                   contribution right here, we have to take it "as is" and pass 
+                   it to the copy data routine. */
+                Tensor<1, dim> normal;
+                if (id == Boundary::slip)
```

Comment:
it is not clear to me what you do for other boundary conditions and if you need to do the work below.

Reply:
We have added a discussion in the introduction outlining how different boundary conditions are implemented in the context of the scheme.